tinygrad/runtime/lib.py

```diff
@@ -109,3 +113,36 @@ def _underlying_buf_memsz(self, buf): return self.buffer_info[buf][0] * self.buf
   def _cached_bufkey(self, size, dtype, device) -> Tuple[int, ...]: return (device, size, dtype, dtype.shape) if isinstance(dtype, ImageDType) else (device, size, dtype) # Provides a key for reusing device buffers with identical keys.
   def _do_alloc(self, size, dtype, device, **kwargs): raise NotImplementedError("must be implemented")
   def _do_free(self, buf): pass
+
+class CachedProgram(abc.ABC):
```

Comment:
Err, not like this. Make a decorator for the compile function, don't extend the Program class